li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java

```diff
@@ -372,17 +439,60 @@ public Long safeOffset(TopicPartition tp) {
 
   @Override
   public void close() {
-    throw new UnsupportedOperationException("Not implemented yet");
+    close(Integer.MAX_VALUE, TimeUnit.MILLISECONDS);
   }
 
   @Override
   public void close(long timeout, TimeUnit timeUnit) {
-    throw new UnsupportedOperationException("Not implemented yet");
+    // Get an immutable copy of the current set of consumers.
+    List<Map.Entry<ClusterDescriptor, LiKafkaConsumer<K, V>>> consumers = _consumers;
+
+    if (consumers.isEmpty()) {
+      LOG.info("no consumers to close for cluster group {}", _clusterGroup);
+      return;
+    }
+
+    LOG.info("closing LiKafkaConsumer for cluster group {} in {} {}...", _clusterGroup, timeout, timeUnit);
+
+    long startTimeMs = System.currentTimeMillis();
+    long deadlineTimeMs = startTimeMs + timeUnit.toMillis(timeout);
+    CountDownLatch countDownLatch = new CountDownLatch(consumers.size());
+    for (Map.Entry<ClusterDescriptor, LiKafkaConsumer<K, V>> entry : consumers) {
+      ClusterDescriptor cluster = entry.getKey();
+      LiKafkaConsumer<K, V> consumer = entry.getValue();
+      Thread t = new Thread(() -> {
+        try {
+          consumer.close(deadlineTimeMs - System.currentTimeMillis(), TimeUnit.MILLISECONDS);
+        } finally {
+          countDownLatch.countDown();
+        }
+      });
+      t.setDaemon(true);
+      t.setName("LiKafkaConsumer-close-" + cluster.name());
+      t.setUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler() {
+        public void uncaughtException(Thread t, Throwable e) {
+          throw new KafkaException("Thread " + t.getName() + " throws exception", e);
+        }
+      });
+      t.start();
+    }
+
+    try {
+      if (!countDownLatch.await(deadlineTimeMs - System.currentTimeMillis(), TimeUnit.MILLISECONDS)) {
+        throw new KafkaException("Fail to close all consumers for cluster group " + _clusterGroup + " in " +
+            timeout + " " + timeUnit);
+      }
+    } catch (InterruptedException e) {
+      throw new KafkaException("Fail to close all consumers for cluster group " + _clusterGroup, e);
+    }
+
+    LOG.info("LiKafkaConsumer close for cluster group {} complete in {} milliseconds", _clusterGroup,
```

Comment:
both the start of the close and the end (here) are info. i feel thats too verbose. maybe make thetop one debug and leave this at info ?

Reply:
done